# Pandas Exploring Data by Filtering and Grouping

## Let's remember cars dataset:
The dataset contains the information of 10.000 cars. There are 9 different columns:
- Make (Car brand, example: Ford)
- Model (The Model of the Car, example: Focus)
- Year (The Year in which the car was build, example: 2012)
- Variant (The car model version showing the PS, example: 1.6 Trendline)
- Kms (The kilometers the cars has been driven, example: 90000)
- Price (The offered price for the car, example: 10000)
- Doors (How many doors the car has, example: 4)
- Kind (Type of car, example: Pick-Up)
- Location (Where the car is located, example: Buenos Aires)



## Recall Dataset

In [ ]:
import pandas as pd

In [ ]:
#reading data
cars = pd.read_csv("https://raw.githubusercontent.com/juliandnl/redi_ss20/master/cars.csv")

In [ ]:
cars.head()

Make   Model  Year  ... Doors       Kind           Location
0  Volkswagen   Vento  2012  ...   4.0      Sedán            Córdoba
1        Ford  Ranger  2012  ...   2.0    Pick-Up         Entre Ríos
2  Volkswagen     Fox  2011  ...   5.0  Hatchback  Bs.as. G.b.a. Sur
3        Ford  Ranger  2017  ...   4.0    Pick-Up            Neuquén
4  Volkswagen     Gol  2013  ...   3.0  Hatchback            Córdoba

[5 rows x 9 columns]

In [ ]:
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Make      10000 non-null  object 
 1   Model     10000 non-null  object 
 2   Year      10000 non-null  int64  
 3   Variant   10000 non-null  object 
 4   Kms       10000 non-null  int64  
 5   Price     10000 non-null  int64  
 6   Doors     10000 non-null  float64
 7   Kind      10000 non-null  object 
 8   Location  10000 non-null  object 
dtypes: float64(1), int64(3), object(5)
memory usage: 703.2+ KB


In [ ]:
cars.nunique()

Make           5
Model         62
Year           8
Variant      592
Kms         1201
Price        948
Doors          4
Kind          10
Location      28
dtype: int64


## Task
1. Checkars is particularly interested in cars of the make Ford from produced in or after 2016. How many Ford cars from 2016 are in the dataset?
2. We would like to analyze cars based on price and km by brands. How expensive are the car makes in general? Which car make is the cheapest one? Which car make has on average the most kms?
3. What are the average prices by brands and by the number of doors that cars have?
4. How many unique models does each brand have?
5. Which kind of cars is the most expensive and most popular?

Bonus:
6. What are the min, max and average production years for Ford cars?


## 1.How many Fords from 2016 are in the dataset?

When you first start to work with a dataset, it's crucial to explore and understand data. During this exploration, you might want to check a subset of data.



For example, you would like to have a subset for just 4 columns namely, Make, Model, Year, Location

In [ ]:
cars[['Make', 'Model', 'Year', 'Location']].head(5)

Make   Model  Year           Location
0  Volkswagen   Vento  2012            Córdoba
1        Ford  Ranger  2012         Entre Ríos
2  Volkswagen     Fox  2011  Bs.as. G.b.a. Sur
3        Ford  Ranger  2017            Neuquén
4  Volkswagen     Gol  2013            Córdoba

There are several ways to filter data from a dataframe in Pandas. We'll cover the 3 of them.


1. Boolean / Dataframe


In [ ]:
cars[cars['Year'] >= 2016]

Make      Model  Year  ... Doors       Kind               Location
3           Ford     Ranger  2017  ...   4.0    Pick-Up                Neuquén
8           Ford  Focus III  2017  ...   5.0  Hatchback  Buenos Aires Interior
29    Volkswagen    Saveiro  2017  ...   2.0    Pick-Up                Córdoba
35          Ford  Focus III  2016  ...   4.0      Sedán      Bs.as. G.b.a. Sur
49          Ford  Focus III  2016  ...   5.0  Hatchback      Bs.as. G.b.a. Sur
...          ...        ...   ...  ...   ...        ...                    ...
9985  Volkswagen  Gol Trend  2016  ...   5.0  Hatchback  Buenos Aires Interior
9990        Ford  Focus III  2016  ...   5.0  Hatchback      Bs.as. G.b.a. Sur
9991        Ford     Ranger  2016  ...   4.0    Pick-Up    Bs.as. G.b.a. Oeste
9994        Ford  Focus III  2016  ...   5.0  Hatchback        Capital Federal
9995        Ford  Focus III  2016  ...   5.0  Hatchback    Bs.as. G.b.a. Oeste

[1414 rows x 9 columns]

In [ ]:
cars[cars.Year >= 2016]

Make      Model  Year  ... Doors       Kind               Location
3           Ford     Ranger  2017  ...   4.0    Pick-Up                Neuquén
8           Ford  Focus III  2017  ...   5.0  Hatchback  Buenos Aires Interior
29    Volkswagen    Saveiro  2017  ...   2.0    Pick-Up                Córdoba
35          Ford  Focus III  2016  ...   4.0      Sedán      Bs.as. G.b.a. Sur
49          Ford  Focus III  2016  ...   5.0  Hatchback      Bs.as. G.b.a. Sur
...          ...        ...   ...  ...   ...        ...                    ...
9985  Volkswagen  Gol Trend  2016  ...   5.0  Hatchback  Buenos Aires Interior
9990        Ford  Focus III  2016  ...   5.0  Hatchback      Bs.as. G.b.a. Sur
9991        Ford     Ranger  2016  ...   4.0    Pick-Up    Bs.as. G.b.a. Oeste
9994        Ford  Focus III  2016  ...   5.0  Hatchback        Capital Federal
9995        Ford  Focus III  2016  ...   5.0  Hatchback    Bs.as. G.b.a. Oeste

[1414 rows x 9 columns]

In [ ]:
cars[cars['Make'] == 'Ford']

Make                  Model  ...       Kind               Location
1     Ford                 Ranger  ...    Pick-Up             Entre Ríos
3     Ford                 Ranger  ...    Pick-Up                Neuquén
6     Ford                     Ka  ...  Hatchback      Bs.as. G.b.a. Sur
7     Ford                     Ka  ...  Hatchback  Buenos Aires Interior
8     Ford              Focus III  ...  Hatchback  Buenos Aires Interior
...    ...                    ...  ...        ...                    ...
9994  Ford              Focus III  ...  Hatchback        Capital Federal
9995  Ford              Focus III  ...  Hatchback    Bs.as. G.b.a. Oeste
9997  Ford  Fiesta Kinetic Design  ...  Hatchback                Tucumán
9998  Ford  Fiesta Kinetic Design  ...  Hatchback  Buenos Aires Interior
9999  Ford              Focus III  ...  Hatchback      Bs.as. G.b.a. Sur

[4025 rows x 9 columns]

In [ ]:
cars[(cars['Make'] == 'Ford') & (cars['Year'] >= 2016)]

Make      Model  Year  ... Doors       Kind               Location
3     Ford     Ranger  2017  ...   4.0    Pick-Up                Neuquén
8     Ford  Focus III  2017  ...   5.0  Hatchback  Buenos Aires Interior
35    Ford  Focus III  2016  ...   4.0      Sedán      Bs.as. G.b.a. Sur
49    Ford  Focus III  2016  ...   5.0  Hatchback      Bs.as. G.b.a. Sur
57    Ford     Mondeo  2018  ...   4.0      Sedán             Entre Ríos
...    ...        ...   ...  ...   ...        ...                    ...
9949  Ford  Focus III  2017  ...   5.0  Hatchback    Bs.as. G.b.a. Oeste
9990  Ford  Focus III  2016  ...   5.0  Hatchback      Bs.as. G.b.a. Sur
9991  Ford     Ranger  2016  ...   4.0    Pick-Up    Bs.as. G.b.a. Oeste
9994  Ford  Focus III  2016  ...   5.0  Hatchback        Capital Federal
9995  Ford  Focus III  2016  ...   5.0  Hatchback    Bs.as. G.b.a. Oeste

[631 rows x 9 columns]

2. `query()`


In [ ]:
cars.query('Make == "Ford"')


Make                  Model  ...       Kind               Location
1     Ford                 Ranger  ...    Pick-Up             Entre Ríos
3     Ford                 Ranger  ...    Pick-Up                Neuquén
6     Ford                     Ka  ...  Hatchback      Bs.as. G.b.a. Sur
7     Ford                     Ka  ...  Hatchback  Buenos Aires Interior
8     Ford              Focus III  ...  Hatchback  Buenos Aires Interior
...    ...                    ...  ...        ...                    ...
9994  Ford              Focus III  ...  Hatchback        Capital Federal
9995  Ford              Focus III  ...  Hatchback    Bs.as. G.b.a. Oeste
9997  Ford  Fiesta Kinetic Design  ...  Hatchback                Tucumán
9998  Ford  Fiesta Kinetic Design  ...  Hatchback  Buenos Aires Interior
9999  Ford              Focus III  ...  Hatchback      Bs.as. G.b.a. Sur

[4025 rows x 9 columns]

In [ ]:
cars.query('Make == "Ford"& Year >=2016')

Make      Model  Year  ... Doors       Kind               Location
3     Ford     Ranger  2017  ...   4.0    Pick-Up                Neuquén
8     Ford  Focus III  2017  ...   5.0  Hatchback  Buenos Aires Interior
35    Ford  Focus III  2016  ...   4.0      Sedán      Bs.as. G.b.a. Sur
49    Ford  Focus III  2016  ...   5.0  Hatchback      Bs.as. G.b.a. Sur
57    Ford     Mondeo  2018  ...   4.0      Sedán             Entre Ríos
...    ...        ...   ...  ...   ...        ...                    ...
9949  Ford  Focus III  2017  ...   5.0  Hatchback    Bs.as. G.b.a. Oeste
9990  Ford  Focus III  2016  ...   5.0  Hatchback      Bs.as. G.b.a. Sur
9991  Ford     Ranger  2016  ...   4.0    Pick-Up    Bs.as. G.b.a. Oeste
9994  Ford  Focus III  2016  ...   5.0  Hatchback        Capital Federal
9995  Ford  Focus III  2016  ...   5.0  Hatchback    Bs.as. G.b.a. Oeste

[631 rows x 9 columns]

3. `loc()` 





In [ ]:
cars.loc[cars['Year'] >= 2016]

Make      Model  Year  ... Doors       Kind               Location
3           Ford     Ranger  2017  ...   4.0    Pick-Up                Neuquén
8           Ford  Focus III  2017  ...   5.0  Hatchback  Buenos Aires Interior
29    Volkswagen    Saveiro  2017  ...   2.0    Pick-Up                Córdoba
35          Ford  Focus III  2016  ...   4.0      Sedán      Bs.as. G.b.a. Sur
49          Ford  Focus III  2016  ...   5.0  Hatchback      Bs.as. G.b.a. Sur
...          ...        ...   ...  ...   ...        ...                    ...
9985  Volkswagen  Gol Trend  2016  ...   5.0  Hatchback  Buenos Aires Interior
9990        Ford  Focus III  2016  ...   5.0  Hatchback      Bs.as. G.b.a. Sur
9991        Ford     Ranger  2016  ...   4.0    Pick-Up    Bs.as. G.b.a. Oeste
9994        Ford  Focus III  2016  ...   5.0  Hatchback        Capital Federal
9995        Ford  Focus III  2016  ...   5.0  Hatchback    Bs.as. G.b.a. Oeste

[1414 rows x 9 columns]

In [ ]:
cars.loc[(cars['Make'] == 'Ford')&(cars['Year'] >=2016)]

Make      Model  Year  ... Doors       Kind               Location
3     Ford     Ranger  2017  ...   4.0    Pick-Up                Neuquén
8     Ford  Focus III  2017  ...   5.0  Hatchback  Buenos Aires Interior
35    Ford  Focus III  2016  ...   4.0      Sedán      Bs.as. G.b.a. Sur
49    Ford  Focus III  2016  ...   5.0  Hatchback      Bs.as. G.b.a. Sur
57    Ford     Mondeo  2018  ...   4.0      Sedán             Entre Ríos
...    ...        ...   ...  ...   ...        ...                    ...
9949  Ford  Focus III  2017  ...   5.0  Hatchback    Bs.as. G.b.a. Oeste
9990  Ford  Focus III  2016  ...   5.0  Hatchback      Bs.as. G.b.a. Sur
9991  Ford     Ranger  2016  ...   4.0    Pick-Up    Bs.as. G.b.a. Oeste
9994  Ford  Focus III  2016  ...   5.0  Hatchback        Capital Federal
9995  Ford  Focus III  2016  ...   5.0  Hatchback    Bs.as. G.b.a. Oeste

[631 rows x 9 columns]

In [ ]:
cars.loc[(cars['Year'] <=2018)&(cars['Year'] >=2016)]

Make      Model  Year  ... Doors       Kind               Location
3           Ford     Ranger  2017  ...   4.0    Pick-Up                Neuquén
8           Ford  Focus III  2017  ...   5.0  Hatchback  Buenos Aires Interior
29    Volkswagen    Saveiro  2017  ...   2.0    Pick-Up                Córdoba
35          Ford  Focus III  2016  ...   4.0      Sedán      Bs.as. G.b.a. Sur
49          Ford  Focus III  2016  ...   5.0  Hatchback      Bs.as. G.b.a. Sur
...          ...        ...   ...  ...   ...        ...                    ...
9985  Volkswagen  Gol Trend  2016  ...   5.0  Hatchback  Buenos Aires Interior
9990        Ford  Focus III  2016  ...   5.0  Hatchback      Bs.as. G.b.a. Sur
9991        Ford     Ranger  2016  ...   4.0    Pick-Up    Bs.as. G.b.a. Oeste
9994        Ford  Focus III  2016  ...   5.0  Hatchback        Capital Federal
9995        Ford  Focus III  2016  ...   5.0  Hatchback    Bs.as. G.b.a. Oeste

[1414 rows x 9 columns]

In [ ]:
cars.loc[(cars['Year'] <=2018) | (cars['Year'] >=2016)]

Make                  Model  ...       Kind               Location
0     Volkswagen                  Vento  ...      Sedán                Córdoba
1           Ford                 Ranger  ...    Pick-Up             Entre Ríos
2     Volkswagen                    Fox  ...  Hatchback      Bs.as. G.b.a. Sur
3           Ford                 Ranger  ...    Pick-Up                Neuquén
4     Volkswagen                    Gol  ...  Hatchback                Córdoba
...          ...                    ...  ...        ...                    ...
9995        Ford              Focus III  ...  Hatchback    Bs.as. G.b.a. Oeste
9996  Volkswagen                   Bora  ...      Sedán  Buenos Aires Interior
9997        Ford  Fiesta Kinetic Design  ...  Hatchback                Tucumán
9998        Ford  Fiesta Kinetic Design  ...  Hatchback  Buenos Aires Interior
9999        Ford              Focus III  ...  Hatchback      Bs.as. G.b.a. Sur

[10000 rows x 9 columns]

In [ ]:
cars['Year'] <=2018

0       True
1       True
2       True
3       True
4       True
        ... 
9995    True
9996    True
9997    True
9998    True
9999    True
Name: Year, Length: 10000, dtype: bool

### Conclusion: 
There are 631 Ford cars produced in or after 2016 in our dataset.

## 2.Let's analyze cars based on price and kms by brands. 
 

1. How expensive are the car makes in general?
2. Which car make is the cheapest one? 
3. Which car make has on average the most Kms?

The `groupby` function is one of the most used function of the pandas library. In addition to filtering, grouping data during the exploration is quite useful.

In [ ]:
cars.groupby('Make')[['Price']].mean()
#cars.groupby('Make')[['Price']].mean().round(1).sort_values('Price', ascending= False)

Price
Make                        
Chrysler       366845.833333
Ford           353041.877764
Honda          381520.603448
Mercedes Benz  512749.515625
Volkswagen     335441.534119

We can also add another column to the groups. For example, what is the average kms by each make?

In [ ]:
cars.groupby('Make')[['Price','Kms']].mean().round(1)

Price      Kms
Make                            
Chrysler       366845.8  79594.0
Ford           353041.9  69897.5
Honda          381520.6  81619.7
Mercedes Benz  512749.5  67358.7
Volkswagen     335441.5  77660.2

### Conclusion:
- The most expensive car make is Mercedes Benz
- Volkswagen is the cheapest car make.
- The most used cars in Honda make (The highest km)

## 3.What are the average prices by brands and by the number of doors that cars have?

In [ ]:
cars.groupby(['Make','Doors'])[['Price']].mean().round(1)

Price
Make          Doors          
Chrysler      5.0    366845.8
Ford          2.0    481503.9
              3.0    158703.6
              4.0    455984.9
              5.0    331894.6
Honda         3.0     26000.0
              4.0    323781.2
              5.0    431716.5
Mercedes Benz 2.0    574315.4
              4.0    497913.1
              5.0    476930.1
Volkswagen    2.0    379124.7
              3.0    226298.6
              4.0    422831.0
              5.0    271274.8

In [ ]:
cars[['Make']]

Make
0     Volkswagen
1           Ford
2     Volkswagen
3           Ford
4     Volkswagen
...          ...
9995        Ford
9996  Volkswagen
9997        Ford
9998        Ford
9999        Ford

[10000 rows x 1 columns]

### Conclusion:
- For Volkswagen, the most expensive cars have 4 doors.
- For Mercedes, the most expensive cars have 2 doors.

## 4.How many unique models does have each brand?

In [ ]:
cars.groupby('Make')[['Model']].nunique()

Model
Make                
Chrysler           2
Ford              11
Honda              6
Mercedes Benz     15
Volkswagen        28

### Conclusion
- There are 28 different models in Volkswagen.
- The least variety is in Chrysler: only 2 models.

## 5.Which kind of cars is the most expensive and most popular?

The `agg()` method allows you to perform specific aggregation(s) on specific columns.

In [ ]:
cars.groupby('Kind')[['Price']].agg(['mean','count']).round(1)

Price      
                 mean count
Kind                       
Cabriolet     79272.6    11
Coupé        543092.2    90
Furgón       702054.7    53
Hatchback    276554.8  4056
Minivan      650000.0     1
Monovolumen  270500.2   705
Pick-Up      545194.6  1547
Rural        330247.1    59
SUV          380708.0  1305
Sedán        338569.6  2173

### Conclusion
- The most popular kind is Hatchback and then Sedan.
- The most expensive kind is Furgon, while the cheapest one is Cabriolet.

## BONUS: What are the min, max and average production year for Ford cars?

In [ ]:
ford_cars = cars[cars['Make'] == 'Ford']
ford_cars['Year'].agg(['mean','min','max']).round(1)

mean    2013.5
min     2011.0
max     2018.0
Name: Year, dtype: float64

In [ ]:
#cars[cars['Make'] == 'Ford']['Year'].agg(['mean','min','max']).round(1)

## Further Materials

Useful aggregation functions:
- `mean()`
- `median()`
- `var()`
- `std()`
- `min()`
- `max()`
- `count()`
- `sum()`
- `nunique()`

### Sorting Data

Let's say, we would like to see the least used cars (in terms of kms).

We can use the `sort_values()` for this purpose. (`sort_values()` sorts ascending by deafult.)



In [ ]:
cars.sort_values('Kms').head(5)

NameError: ignored

We can modify the order of sort_values if we would like to see the most used cars.

In [ ]:
cars.sort_values('Kms', ascending=False).head(5)

NameError: ignored

**NOTE**: Using the `sort_values()` method does not overwrite the object. It remains unaltered after applying it the `sort_values()` function

If you want to overwrite the object with the new values ordered there are 2 ways to do it:


1.   Assign to `cars` the value `cars.sort_values()`




In [ ]:
sorted_cars = cars.sort_values('Kms')
sorted_cars.head(5)

2.   Use the `inplace` `parameter` of `.sort_values()`

In [ ]:
cars.sort_values('Kms', ascending=False, inplace = True)
cars.head(5)

Another way to sort data in Pandas is using sort_index(). As its name suggests this statement is used for sorting indexes. 

In [ ]:
cars.sort_index()

In [ ]:
cars.sort_index(inplace= True) 
#cars.sort_index(ascending = False)
cars

### Exercise
Find which **10 cars** has the highest prices. Please use `sort_values()` and `head()`.
While finding the top ten expensive cars overwrite the dataset and then return back to original sorting by using `sort_index()`.

In [ ]:
cars.head()

In [ ]:
cars.sort_values('Price', ascending=False).head(10)

In [ ]:
cars.sort_values('Price', ascending=False, inplace = True) 
#or cars = cars.sort_values('Price', ascending=False).head(10)
cars.head(10)